subsys/net/lib/http/http_app.c

```diff
@@ -0,0 +1,247 @@
+#include <string.h>
+#include <strings.h>
+#include <errno.h>
+#include <stdlib.h>
+#include <version.h>
+
+#include <net/net_core.h>
+#include <net/net_ip.h>
+#include <net/http.h>
+
+int http_set_cb(struct http_ctx *ctx,
+		http_connect_cb_t connect_cb,
+		http_recv_cb_t recv_cb,
+		http_send_cb_t send_cb,
+		http_close_cb_t close_cb)
+{
+	if (!ctx) {
+		return -EINVAL;
+	}
+
+	if (!ctx->is_init) {
+		return -ENOENT;
+	}
+
+	ctx->cb.connect = connect_cb;
+	ctx->cb.recv = recv_cb;
+	ctx->cb.send = send_cb;
+	ctx->cb.close = close_cb;
+
+	return 0;
+}
+
+int http_close(struct http_ctx *ctx)
+{
+	if (!ctx) {
+		return -EINVAL;
+	}
+
+	if (!ctx->is_init) {
+		return -ENOENT;
+	}
+
+#if defined(CONFIG_HTTP_SERVER) && defined(CONFIG_NET_DEBUG_HTTP_CONN)
+	if (!ctx->is_client) {
+		http_server_conn_del(ctx);
+	}
+#endif
+
+	if (ctx->pending) {
+		net_pkt_unref(ctx->pending);
+		ctx->pending = NULL;
+	}
+
+	return net_app_close(&ctx->app_ctx);
+}
+
+int http_release(struct http_ctx *ctx)
+{
+	if (!ctx) {
+		return -EINVAL;
+	}
+
+	if (!ctx->is_init) {
+		return -ENOENT;
+	}
+
+	ctx->is_tls = false;
+
+#if defined(CONFIG_HTTP_SERVER) && defined(CONFIG_NET_DEBUG_HTTP_CONN)
+	if (!ctx->is_client) {
+		http_server_conn_del(ctx);
+		http_server_disable(ctx);
+	}
+#endif
+
+	if (ctx->pending) {
+		net_pkt_unref(ctx->pending);
+		ctx->pending = NULL;
+	}
+
+	ctx->is_init = false;
+
+	return net_app_release(&ctx->app_ctx);
+}
+
+int http_send_msg_raw(struct http_ctx *ctx, struct net_pkt *pkt,
+		      void *user_send_data)
+{
+	int ret;
+
+	NET_DBG("[%p] Sending %zd bytes data", ctx, net_pkt_get_len(pkt));
+
+	ret = net_app_send_pkt(&ctx->app_ctx, pkt, NULL, 0, 0,
+			       user_send_data);
+	if (!ret) {
+		/* We must let the system to send the packet, otherwise TCP
+		 * might timeout before the packet is actually sent. This is
+		 * easily seen if the application calls this functions many
+		 * times in a row.
+		 */
+		k_yield();
+	}
+
+	return ret;
+}
+
+static int _http_add_chunk(struct net_pkt *pkt, s32_t timeout,
+			   const u8_t *buf, size_t len)
+{
+	char chunk_header[16];
+	size_t ret;
+
+	if (!buf) {
+		len = 0;
+	}
+
+	snprintk(chunk_header, sizeof(chunk_header), "%x" HTTP_CRLF,
+		 (unsigned int)len);
+
+	ret = net_pkt_append(pkt, strlen(chunk_header), chunk_header,
+			     timeout);
+	if (ret == 0) {
+		return 0;
+	}
+
+	if (len) {
+		len = net_pkt_append(pkt, len, buf, timeout);
+		if (len == 0) {
+			return 0;
+		}
+	}
+
+	ret = net_pkt_append(pkt, sizeof(HTTP_CRLF) - 1, HTTP_CRLF,
+			     timeout);
+	if (ret == 0) {
+		return 0;
+	}
+
+	return len;
+}
+
+int http_send_data(struct http_ctx *ctx,
+		   const char *payload,
+		   size_t payload_len,
+		   void *user_send_data)
+{
+	size_t added;
+	int ret;
+
+	if (!ctx->pending) {
+		ctx->pending = net_app_get_net_pkt(&ctx->app_ctx,
+						   AF_UNSPEC,
+						   ctx->timeout);
+		if (!ctx->pending) {
+			return -ENOMEM;
+		}
+	}
+
+	ret = _http_add_chunk(ctx->pending, ctx->timeout,
```

Comment:
Well, so the idea was that http_send_data() is the most basic, primitive function to send data over HTTP connection. All other sending is done solely via it. Sending a transfer-eccoding-chunked chunk is thus a higher-level operation, with implementation like:

~~~
add_chunk(ctx, buf, len)
{
  char sizebuf[10];
  int sz = sprintf(sizebuf, "%x\r\n", len);
  http_send_data(ctx, sizebuf, sz); 
  http_send_data(ctx, buf, len); 
  http_send_data(ctx, "\r\b", 2); 
}
~~~

Reply:
I disagree here. IMHO the lowest function should be the one that is using the net_buf as a parameter. In this case it would be http_send_msg_raw(). If developer knows what he/she is doing, then that function could be used to send data.

Reply:
> If developer knows what he/she is doing, then that function could be used to send data.

That statement is always true, and I always agree with it. The point, in the paradigm I propose, such a function will never be needed by an end user, and the only use will be from `http_send_bytes()`, and all other funcs will use `http_send_bytes()` (and never `http_send_msg_raw()` directly). As long as we on the same line with that, we're talking about the same paradigm, otherwise, we're still talking about different things.

Reply:
Sure, I agree.